### PR TITLE
Discovery 4.1 docs

### PR DESCRIPTION
**Pull request by lzap**
DO NOT MERGE, this is still WIP.

Sachin, here are the ideal workflows to test:
- [ ] Traditional discovery
- [ ] Traditional discovery with auto-provisioning
- [ ] ISO with guest with single NIC on a network without DHCP or PXE
- [ ] ISO with guest with single NIC on a network with DHCP but without PXE
- [ ] ISO with guest with single NIC on a network with DHCP and PXE
- [ ] ISO with guest with single NIC on a network without DHCP or PXE, semi-automated
- [ ] ISO with guest with single NIC on a network with DHCP but without PXE, semi-automated
- [ ] ISO with guest with single NIC on a network with DHCP and PXE, semi-automated
- [ ] ISO with guest with single NIC on a network without DHCP or PXE, unattended
- [ ] ISO with guest with single NIC on a network with DHCP but without PXE, unattended
- [ ] ISO with guest with single NIC on a network with DHCP and PXE, unattended
- [ ] ISO with guest with multiple NIC on a network with DHCP but without PXE
- [ ] ISO with guest with multiple NIC on a network with DHCP and PXE
- [ ] ISO with guest with multiple NIC on a network without DHCP or PXE, semi-automated
- [ ] ISO with guest with multiple NIC on a network with DHCP but without PXE, semi-automated
- [ ] ISO with guest with multiple NIC on a network with DHCP and PXE, semi-automated
- [ ] ISO with guest with multiple NIC on a network without DHCP or PXE, unattended
- [ ] ISO with guest with multiple NIC on a network with DHCP but without PXE, unattended
- [ ] ISO with guest with multiple NIC on a network with DHCP and PXE, unattended
- [ ] Boot unmodified ISO on EFI system
- [ ] Boot remastered ISO on EFI system

### CODE REVIEW

File: plugins/foreman_discovery/4.0/index.md [hunk 1]
```diff
@@ -225,12 +225,11 @@ on Debian and Ubuntu systems.
...
 
-## 2.2 Foreman Proxy Discovery plugin
+## 2.2 Foreman Proxy Discovery plugin (optional)
 
-Foreman Proxy Discovery plugin must be installed on all proxy hosts on the
-network where discovery feature is supposed to work. It is possible to skip
-this step if direct communication is planned, but configuration must be
-changed in that case.
+Foreman Proxy plugin must be only installed when Foreman server does not have
+direct connection to. Foreman Proxy Discovery plugin must be installed on all
```
Comment (stbenjam) [inline plugins/foreman_discovery/4.0/index.md]: > Foreman Proxy plugin must be only installed when Foreman server does not have direct connection to. 

Maybe add "to the discovered host."  

Or rephrase, "Smart Proxy plugin can be used to discover hosts in networks Foreman can't directly reach itself."

File: plugins/foreman_discovery/4.0/index.md [hunk 2]
```diff
@@ -225,12 +225,11 @@ on Debian and Ubuntu systems.
 
 Restart of Foreman is needed when installing the package separately.
 
-## 2.2 Foreman Proxy Discovery plugin
+## 2.2 Foreman Proxy Discovery plugin (optional)
```
Comment (stbenjam) [inline plugins/foreman_discovery/4.0/index.md]: `s/Foreman Proxy/Smart Proxy/g`? Most manuals seem to use Smart Proxy.

File: plugins/foreman_discovery/4.0/index.md [hunk 3]
```diff
@@ -707,6 +695,59 @@ foreman-discovery-image git
...
 the README for further instructions.
 
+## 5.3 PXE-less discovery
+
+It is possible to use Discovery image directly as a CDROM/DVDROM ISO that can
+be also transferred to hard drive or USB stick (dd command). In this case,
+automatic discovery process is not started and user select primary network
+interface interactively. On the couple of screens user configures network
+credentials (IPv4 supported at the moment), Foreman URL and type and fills user
+facts which are then uploaded to the server.
```
Comment (stbenjam) [inline plugins/foreman_discovery/4.0/index.md]: Some minor corrections:

It is possible to use **the** Discovery image directly as a CDROM/DVDROM ISO that can
be also transferred to hard drive or USB stick (dd command). In this case,
automatic discovery process is not started and **the** user **can** select **the** primary network
interface interactively. On the **next** couple of screens**, the** user configures network
credentials (IPv4 supported at the moment), **the** ~~Foreman~~ URL and type **(Foreman or Smart Proxy)**, and fills **custom** user facts which are then uploaded to the server.

Comment (stbenjam) [inline plugins/foreman_discovery/4.0/index.md]: Maybe also discuss the custom facts more.  "These custom facts can be used with pre-defined discovery rules to let users automatically build hosts."  And then include the authentication comment here

File: plugins/foreman_discovery/4.0/index.md [hunk 4]
```diff
@@ -707,6 +695,59 @@ foreman-discovery-image git
...
+a signal to Discovery plugin to initiate kexec reload instead of reboot. When
+the host is provisioned either manually or via discovery rule, instance loads
+installer via kexec utility.
+
+This interactive mode is also available when image is PXE booted, user must
+cancel initial splash screen countdown with a key to get there.
+
+Currently there is authentication in the workflow (the same applies for normal
+discovery process), but it is possible to define own custom facts named "token"
+or "password" to secure PXE-less provisioning.
```
Comment (stbenjam) [inline plugins/foreman_discovery/4.0/index.md]: Maybe move this to the section about about custom facts.

**Comment by stbenjam** [comment]
Otherwise looks good to me!

File: plugins/foreman_discovery/4.0/index.md (hunk 2 above)
Comment (lzap) [inline plugins/foreman_discovery/4.0/index.md]: Replaced all.

File: plugins/foreman_discovery/4.0/index.md (hunk 1 above)
Comment (lzap) [inline plugins/foreman_discovery/4.0/index.md]: Rephrased.

File: plugins/foreman_discovery/4.0/index.md (hunk 3 above)
Comment (lzap) [inline plugins/foreman_discovery/4.0/index.md]: Shame on me, incorporated all your corrections, thanks.

**Comment by lzap** [comment]
Amended all the changes and added example scenarios. Still need to add info how to modify ISO SYSLINUX options.

**Comment by lzap** [comment]
Pls do not merge until Sachin confirms technically.

**Comment by sghai** [comment]
This issue is blocking to test all PXELess scenarios http://projects.theforeman.org/issues/11978

**Comment by lzap** [comment]
I believe you have a hotfix now, how far are you with testing? Can you check those boxes ?

**Comment by sghai** [comment]
Yes. Hotfix is available and testing is in-progress. I can't update the checkboxes in original description. So updating here.. I'll keep this thread updated with progress..
- [x] Traditional discovery
- [x] Traditional discovery with auto-provisioning
- [x] ISO with guest with single NIC on a network without DHCP or PXE
- [x] ISO with guest with single NIC on a network with DHCP but without PXE
- [x] ISO with guest with single NIC on a network with DHCP and PXE
- [x] ISO with guest with single NIC on a network without DHCP or PXE, semi-automated
- [x] ISO with guest with single NIC on a network with DHCP but without PXE, semi-automated
- [x] ISO with guest with single NIC on a network with DHCP and PXE, semi-automated
- [x] ISO with guest with single NIC on a network without DHCP or PXE, unattended
- [ ] ISO with guest with single NIC on a network with DHCP but without PXE, unattended
- [x] ISO with guest with single NIC on a network with DHCP and PXE, unattended
- [x] ISO with guest with multiple NIC on a network with DHCP but without PXE
- [x] ISO with guest with multiple NIC on a network with DHCP and PXE
- [x] ISO with guest with multiple NIC on a network without DHCP or PXE, semi-automated
- [x] ISO with guest with multiple NIC on a network with DHCP but without PXE, semi-automated
- [x] ISO with guest with multiple NIC on a network with DHCP and PXE, semi-automated
- [x] ISO with guest with multiple NIC on a network without DHCP or PXE, unattended
- [ ] ISO with guest with multiple NIC on a network with DHCP but without PXE, unattended
- [x] ISO with guest with multiple NIC on a network with DHCP and PXE, unattended
- [ ] Boot unmodified ISO on EFI system
- [x] Boot remastered ISO on EFI system

**Comment by sghai** [comment]
Couple of things.. 
1) All PXE based scenario tested via PXE booting and for that I used `fdi-bootable-3-0.1.iso` tar ball (not ISO).
2) Currently, I'm facing issue for all PXEless scenario's where discovered host doesn't resolve  foreman_server's hostname and provisioning stuck at URL setup screen. However same works for all PXEbased scenarios.

I'm going to give that a shot today to find the root cause of hostname resolution issue and will file an issue(if needed).

**Comment by sghai** [comment]
Filed issue related to scenario's `network with DHCP but without PXE, unattended`
http://projects.theforeman.org/issues/12069

**Comment by sghai** [comment]
The scenario `Boot unmodified ISO on EFI system` failed at media check. However remaster works fine.. I can see  `bios_vendor   EFI Development Kit II / OVMF` in webUI.

**Comment by lzap** [comment]
Moved into new file, this is now ready for merge, thanks!

**Comment by lzap** [comment]
@stbenjam if you have spare minute, this needs review, thanks!

File: plugins/foreman_discovery/4.1/index.md [hunk 5]
```diff
@@ -0,0 +1,993 @@
...
+    # apt-get install ruby-foreman-discovery
+
+on Debian and Ubuntu systems.
+
+Restart of Foreman is needed when installing the package separately.
+
+## 2.2 Smart Proxy Discovery plugin (optional)
+
+Smart Proxy plugin can be used to discover hosts in networks Foreman can't
+directly reach itself. The plugin plugin must be installed on all proxy hosts
```
Comment (stbenjam) [inline plugins/foreman_discovery/4.1/index.md]: plugin plugin

File: plugins/foreman_discovery/4.1/index.md [hunk 6]
```diff
@@ -0,0 +1,993 @@
...
+
+Once the template has been updated, click the *Build PXE Default* button at the
+top of the Provisioning Templates page. This will instruct the TFTP proxy to
+rewrite the *pxelinux.cfg/default* file.
+
+Repeat this step every time a change is made to the default template.
+
+## 3.1.3 Organizations and Locations
+
+If Locations and/or Organisations are enabled, Foreman assigns Organization
```
Comment (stbenjam) [inline plugins/foreman_discovery/4.1/index.md]: Use American "Organizations" spelling for consistency

**Comment by stbenjam** [comment]
@lzap Just those two little things, then good to go for me

**Comment by lzap** [comment]
Amended, thanks!

File: plugins/foreman_discovery/4.1/index.md [hunk 7]
```diff
@@ -0,0 +1,993 @@
...
+
+Important note: Executing foreman-installer will re-deploy all foreman-related
+configuration files. In case some changes have been made, proceed with manual
+download described below.
+
+### 2.1.2 Manual installation
+
+To install Foreman Discovery plugin, do the following
+
+    # yum install ruby193-rubygem-foreman_discovery
```
Comment (domcleal) [inline plugins/foreman_discovery/4.1/index.md]: `tfm` prefix, if this is for Foreman 1.10.

File: plugins/foreman_discovery/4.1/index.md [hunk 8]
```diff
@@ -0,0 +1,993 @@
...
+## 2.4 Hammer CLI Foreman Discovery
+
+### 2.4.1 Installation
+
+Before installing make sure you have installed the
+[hammer_cli_foreman plugin](https://github.com/theforeman/hammer-cli/blob/master/doc/installation.md),
+
+On Red Hat compatible systems:
+
+```yum install rubygem-hammer_cli_foreman_discovery```
```
Comment (domcleal) [inline plugins/foreman_discovery/4.1/index.md]: `tfm-rubygem-hammer_cli_foreman_discovery` in Foreman 1.10

File: plugins/foreman_discovery/4.1/index.md [hunk 9]
```diff
@@ -0,0 +1,993 @@
...
+    <td>= 1.9</td>
+    <td>4.0.0</td>
+    <td>1.0.2</td>
+    <td>2.1.1</td>
+    <td>0.0.2</td>
+  </tr>
+  <tr>
+    <td>= 1.10</td>
+    <td>4.1.1</td>
+    <td>1.0.5</td>
```
Comment (domcleal) [inline plugins/foreman_discovery/4.1/index.md]: 1.0.3

File: plugins/foreman_discovery/index.md [hunk 10]
```diff
@@ -25,4 +25,10 @@ title: Foreman discovery documentation
 			<p id='manual'>4.0</p>
 		</a>
 	</div>
+  <div class='col-md-4 center'>
+		<a href="plugins/foreman_discovery/4.1/index.html" class="btn-doc btn">
+			<i class="fa fa-newspaper-o"></i>
+			<p id='manual'>4.1</p>
```
Comment (domcleal) [inline plugins/foreman_discovery/index.md]: Nitpick, this page is the opposite order to the Salt page, which goes latest -> oldest.

**Comment by lzap** [comment]
Done!

**Comment by domcleal** [comment]
Thanks!

**Comment by lzap** [comment]
Thank you!

**Pull request merged**